src/provider/provider_tracking.c

```diff
@@ -382,6 +382,12 @@ static umf_result_t trackingFree(void *hProvider, void *ptr, size_t size) {
     ret = umfMemoryProviderFree(p->hUpstream, ptr, size);
     if (ret != UMF_RESULT_SUCCESS) {
         LOG_ERR("upstream provider failed to free the memory");
+        // Do not add memory back to the tracker,
+        // if the provider does not support the free() op.
+        if (ret == UMF_RESULT_ERROR_NOT_SUPPORTED) {
```

Comment:
I agree that we should not add memory back to the tracker in case of memory provider does not support `free`.

But why is pool implementation tries to call `umfMemoryProviderFree()` at all?

Reply:
Do you mean `umfMemoryProviderFree()` above ? - inside `trackingFree()` in line no. 382 ?

Reply:
With such fix we get a memory leak.

Imagine the pool calls `umfMemoryProviderFree()` for the address range `[a. b)`. After that pool calls `umfMemoryProviderAlloc()` my understanding is that memory provider will never return the range `[a. b)` as a result of malloc, but it will return the next chunk `[b,c)`. If the pool will call `umfMemoryProviderFree()` for the `[b,c)` range it also will be lost and never be used again.

Reply:
Yes, the file provider and the devdax provider work exactly as as you wrote - they were designed to work in this way.

Reply:
... but we should not just return "success" here because the memory is actually not released. Maybe this fix is wrong... We should fix only the case when the pool that uses devdax/file providers is destroyed. Only then we should remove the entries from the tracker as we also unmap the memory. 

There is also an additional way fix this bug - free() in devdax/file could do the unmap but we could keep the offset unchanged - this way the free() would be supported and entries would be correctly removed from the tracker while we still keep file/devdax provider logic simple (so the memory would be still unavailable after free).

Reply:
> There is also an additional way fix this bug - free() in devdax/file could do the unmap but we could keep the offset unchanged - this way the free() would be supported and entries would be correctly removed from the tracker while we still keep file/devdax provider logic simple (so the memory would be still unavailable after free).

I agree that at least we are not wasting VA space (but it is not a problem), but since we never decrement the offset in devdax and file providers by design we still have a kind of memory leak. Because if the `umfMemoryProviderAlloc` and `umfMemoryProviderFree` are called multiple times, after some time `umfMemoryProviderAlloc` will return OOM because offset reached the end of file/devdax.


We discussed this with @ldorau offline. My main comment that currently this PR eliminates the issue:
 - today we have a memory leak if `umfMemoryProviderFree` is called. If we do not remove the range from the tracker our diagnostic in DEBUG mode will detect it and report.
 - with this PR the memory leak still there but our diagnostic is not able to detect it and report.

Reply:
> I agree with @ldorau - this is how these providers work. And because of this, in the real-life scenarios they should be used either with a pool that never calls free() or with CoarseProvider

Just to clarify, I am not challenging how devdax or file providers work. we agreed on that. My concern is about this particular flow when some pool calls `umfmemoryProviderFree` for devdax or file providers. This situation should not happens or should be handled correctly. Today looks like there is a memory leak in that case and this PR does not fix the leak.

Reply:
Removed (code change). It will be fixed in another way.
Done.

Reply:
I was saying that providers without free should have a null ptr to function, and we should have an API that pool can check if free function is supported if not - if it's not supported it should never call providerFree(). 

Or as another option - pool informs umf if it can support provider without free - if pool cannot handle it, we always wrap it with coarse provider.

Reply:
Code changed.
Done.